src/mscorlib/shared/System/String.cs

```diff
@@ -2,109 +2,391 @@
-        // Leaving it uninitialized would confuse debuggers.
-        //
-        //We need to call the String constructor so that the compiler doesn't mark this as a literal.
-        //Marking this as a literal would mean that it doesn't show up as a field which we can access 
-        //from native.
-        public static readonly String Empty;
+#if PROJECTN
+        [DependencyReductionRoot]
+#endif
+#if !CORECLR
+        static
+#endif
+        private string Ctor(char[] value)
+        {
+            if (value != null && value.Length != 0)
+            {
+                string result = FastAllocateString(value.Length);
+                unsafe
+                {
+                    fixed (char* dest = &result._firstChar, source = value)
+                    {
+                        wstrcpy(dest, source, value.Length);
+                    }
+                }
+                return result;
+            }
+            else
+                return Empty;
+        }
 
-        internal char FirstChar { get { return _firstChar; } }
-        //
-        // This is a helper method for the security team.  They need to uppercase some strings (guaranteed to be less 
-        // than 0x80) before security is fully initialized.  Without security initialized, we can't grab resources (the nlp's)
-        // from the assembly.  This provides a workaround for that problem and should NOT be used anywhere else.
-        //
-        internal unsafe static string SmallCharToUpper(string strIn)
-        {
-            Debug.Assert(strIn != null);
-            //
-            // Get the length and pointers to each of the buffers.  Walk the length
-            // of the string and copy the characters from the inBuffer to the outBuffer,
-            // capitalizing it if necessary.  We assert that all of our characters are
-            // less than 0x80.
-            //
-            int length = strIn.Length;
-            String strOut = FastAllocateString(length);
-            fixed (char* inBuff = &strIn._firstChar, outBuff = &strOut._firstChar)
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        public extern String(char[] value, int startIndex, int length);
+
+#if PROJECTN
+        [DependencyReductionRoot]
+#endif
+#if !CORECLR
+        static
+#endif
+        private string Ctor(char[] value, int startIndex, int length)
+        {
+            if (value == null)
+                throw new ArgumentNullException(nameof(value));
+
+            if (startIndex < 0)
+                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_StartIndex);
+
+            if (length < 0)
+                throw new ArgumentOutOfRangeException(nameof(length), SR.ArgumentOutOfRange_NegativeLength);
+
+            if (startIndex > value.Length - length)
+                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_Index);
```

Comment:
Any benefit to switching a bunch of these to ThrowHelper?

Reply:
ThrowHelper only helps for generic methods (because of the code size hit is multiplied) or tiny methods that do not call anything else (because of it makes the frame smaller).

ThrowHelper is pain to maintain. Folks tend to be using it more often than necessary. I hope somebody is eventually going to do ILLinker-based optimization to generate it automatically and we will be able to delete all the manually maintained ThrowHelpers.